interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,367 @@
+	// - `\`: also dangerous
+	customDeviceUDevDeviceRegexp = regexp.MustCompile(`^/dev/[^"|{}\\]+$`)
+
+	// Validating regexp for udev tag values (all but kernel devices)
+	customDeviceUDevValueRegexp = regexp.MustCompile(`^[^"{}\\]+$`)
+)
+
+// customDeviceInterface allows sharing customDevice between snaps
+type customDeviceInterface struct{}
+
+func (iface *customDeviceInterface) validateFilePath(path string) error {
+	if !customDevicePathRegexp.MatchString(path) {
+		return fmt.Errorf(`custom-device path must start with / and cannot contain special characters: %q`, path)
+	}
+
+	if !cleanSubPath(path) {
+		return fmt.Errorf(`custom-device path is not clean: %q`, path)
+	}
+
+	if _, err := utils.NewPathPattern(path); err != nil {
+		return fmt.Errorf(`custom-device path cannot be used: %v`, err)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateDevice(path string) error {
+	// The device must satisfy udev's device name rules and generic path rules
+	if !customDeviceUDevDeviceRegexp.MatchString(path) {
+		return fmt.Errorf(`custom-device path must start with /dev/ and cannot contain special characters: %q`, path)
+	}
+
+	if err := iface.validateFilePath(path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validatePaths(slot *snap.SlotInfo, attrName string) error {
+	var paths []string
+	if err := slot.Attr(attrName, &paths); err != nil {
+		if !errors.Is(err, snap.AttributeNotFoundError{}) {
+			return err
+		}
+	}
+
+	for _, path := range paths {
+		if err := iface.validateFilePath(path); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevValue(value interface{}) error {
+	stringValue, ok := value.(string)
+	if !ok {
+		return fmt.Errorf(`value "%v" is not a string`, value)
+	}
+
+	if !customDeviceUDevValueRegexp.MatchString(stringValue) {
+		return fmt.Errorf(`value "%v" contains invalid characters`, stringValue)
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevValueMap(value interface{}) error {
+	valueMap, ok := value.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf(`value "%v" is not a map`, value)
+	}
+
+	for key, val := range valueMap {
+		if !customDeviceUDevValueRegexp.MatchString(key) {
+			return fmt.Errorf(`key "%v" contains invalid characters`, key)
+		}
+		if err := iface.validateUDevValue(val); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) validateUDevTaggingRule(rule map[string]interface{}, devices []string) error {
+	hasKernelTag := false
+	for key, value := range rule {
+		var err error
+		switch key {
+		case "subsystem":
+			err = iface.validateUDevValue(value)
+		case "kernel":
+			hasKernelTag = true
+			err = iface.validateUDevValue(value)
+			if err == nil {
+				deviceName := value.(string)
+				// furthermore, the kernel name must match the name of one of
+				// the given devices
```

Comment:
is this true? cc @anonymouse64

Reply:
My idea is that using `strutil.ListContains()` imposes a restriction that the `devices` and `kernel` udev rule must be an exact match, but globs are still allowed as long as they match exactly. I think it's a restriction good to have, because it makes the store people's lives easier and it does not really harm the user of this interface in any way.